[macOS] Add Android NDK r23

nikita-bykov opened the pull request:
# Description
In scope of this PR we have added Android NDK r23 to macOS images and leave the current LTS version (r21) as default.
Also, we've reworked the `android.ndk` toolset section:
- The `android.ndk.versions` field  is a manually sorted list of all versions that will be installed on the image
- The `android.ndk.lts` field has been renamed to `android.ndk.default`. It will allow us, if necessary, to set a non-LTS NDK version as default.
- The `android.ndk.latest` field has been removed and its value will be taken from the end of `android.ndk.versions` list



#### Related issue: [Update/Add Android NDK r23 LTS #3894](https://github.com/actions/virtual-environments/issues/3894)
## Check list
- [x] Related issue / work item is attached
- [x] Tests are written (if applicable)
- [x] Documentation is updated (if applicable)
- [x] Changes are tested and related VM images are successfully generated: [macOS 10.14](https://github.visualstudio.com/virtual-environments/_build/results?buildId=116323&view=results), [macOS 10.15](https://github.visualstudio.com/virtual-environments/_build/results?buildId=116324&view=results), [macOS 11](https://github.visualstudio.com/virtual-environments/_build/results?buildId=116322&view=results)

maxim-lobanov commented on images/macos/provision/core/android-toolsets.sh:
```diff
@@ -31,8 +31,9 @@ ANDROID_BUILD_TOOL=($(get_toolset_value '.android.build_tools_min_version'))
-ANDROID_NDK_MAJOR_LTS=($(get_toolset_value '.android.ndk.lts'))
-ANDROID_NDK_MAJOR_LATEST=($(get_toolset_value '.android.ndk.latest'))
+ANDROID_NDK_MAJOR_VERSIONS=($(get_toolset_value '.android.ndk."versions"[]'))
+ANDROID_NDK_MAJOR_DEFAULT=($(get_toolset_value '.android.ndk.default'))
```
Why do we need `=(...)` brackets here?

nikita-bykov replied:
Removed unnecessary brackets

maxim-lobanov commented on images/macos/provision/core/android-toolsets.sh:
```diff
@@ -31,8 +31,9 @@ ANDROID_BUILD_TOOL=($(get_toolset_value '.android.build_tools_min_version'))
-ANDROID_NDK_MAJOR_LATEST=($(get_toolset_value '.android.ndk.latest'))
+ANDROID_NDK_MAJOR_VERSIONS=($(get_toolset_value '.android.ndk."versions"[]'))
+ANDROID_NDK_MAJOR_DEFAULT=($(get_toolset_value '.android.ndk.default'))
+ANDROID_NDK_MAJOR_LATEST=(${ANDROID_NDK_MAJOR_VERSIONS[${#ANDROID_NDK_MAJOR_VERSIONS[@]}-1]})
```
nitpick: may be `$(get_toolset_value '.android.ndk."versions"[-1]')` will be more readable?

nikita-bykov replied:
Fixed it

miketimofeev commented:
@nikita-bykov @maxim-lobanov @MaksimZhukov two thoughts:
- Does it make sense to add a test checking that ndk-bundle points to the default ndk version from toolset?
- Shouldn't we add the `(default)` label against the ndk-bundle version in the docs?
![image](https://user-images.githubusercontent.com/48208649/132484842-8bafaa72-4eea-4613-9e31-a487090c3278.png)

al-cheb commented on images/macos/tests/Android.Tests.ps1:
```diff
@@ -8,10 +8,8 @@ Describe "Android" {
-    $ndkLatestFullVersion = (Get-ChildItem "$env:ANDROID_HOME/ndk/$ndkLatestVersion.*" | Select-Object -Last 1).Name
-    $ndkLtsFullVersion = (Get-ChildItem "$env:ANDROID_HOME/ndk/$ndkLtsVersion.*" | Select-Object -Last 1).Name
+    [array]$ndkVersions = Get-ToolsetValue "android.ndk.versions"
+    $ndkFullVersions = $ndkVersions | ForEach-Object { (Get-ChildItem "$env:ANDROID_HOME/ndk/${_}.*" | Select-Object -Last 1).Name }
```
nitpick: 

`$ndkVersions | ForEach-Object { Get-ChildItem "$env:ANDROID_HOME/ndk/${_}.*" -Name | Select-Object -Last 1}`

nikita-bykov replied:
Fixed it

al-cheb commented on images/macos/tests/Android.Tests.ps1:
```diff
@@ -28,10 +26,9 @@ Describe "Android" {
-        "ndk/$ndkLtsFullVersion",
         $platforms,
         $buildTools,
+        ($ndkFullVersions | ForEach-Object { "ndk/${_}" }),
```
nitpick: ($ndkFullVersions | ForEach-Object { "ndk/${_}" }) -> $ndkFullVersions and move `ForEach-Object { "ndk/${_}" ` part at line 12

nikita-bykov replied:
Fixed it

al-cheb commented on images/macos/provision/core/android-toolsets.sh:
```diff
@@ -59,13 +60,17 @@ echo "Installing latest tools & platform tools..."
-ndkLtsLatest=$(get_full_ndk_version  $ANDROID_NDK_MAJOR_LTS)
-ndkLatest=$(get_full_ndk_version  $ANDROID_NDK_MAJOR_LATEST)
-echo y | $SDKMANAGER  "ndk;$ndkLtsLatest" "ndk;$ndkLatest"
+ndkDefault=$(get_full_ndk_version $ANDROID_NDK_MAJOR_DEFAULT)
```
Let's move line 63-64 at line 72

nikita-bykov replied:
Moved it

nikita-bykov commented:
> @nikita-bykov @maxim-lobanov @MaksimZhukov two thoughts:
> 
> * Does it make sense to add a test checking that ndk-bundle points to the default ndk version from toolset?
> * Shouldn't we add the `(default)` label against the ndk-bundle version in the docs?
>   ![image](https://user-images.githubusercontent.com/48208649/132484842-8bafaa72-4eea-4613-9e31-a487090c3278.png)

Added a new test and updated the documentation for the default NDK version
